pkg/dns/plugin.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dns
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	dnsapi "istio.io/istio/pkg/dns/coredns/pb"
+	"istio.io/pkg/env"
+	"istio.io/pkg/log"
+
+	"google.golang.org/grpc"
+)
+
+// Based on istio-ecosystem/istio-coredns-plugin
+// Changes from original:
+// - runs inside istiod, using Istio main gRPC server
+// - instead of directly reading from K8S, use istiod store.
+// - removed "log" - switching to istio log.
+// - refactored Query, so both DNS native interface and coredns grpc plugin are implemented
+// - added parts of istio-ecosystem/dns-discovery, to provide in process DNS
+
+// K8S DNS uses container port 10053, and appears to have a dnsmasq container
+// alongside.
+
+// IstioServiceEntries exposes a DNS interface to internal Istiod service database.
+// This can be used:
+// - as a CoreDNS gRPC plugin
+// - as a DNS-over-TLS resolver, with support for forwarding to k8s or upstream
+// - for debug - a plain DNS-over-UDP interface.
+//
+// The code is currently targeted for Istiod, with a per/pod or per/VM coreDNS
+// forwarding to it, and using the same XDS grpc server and cert.
+//
+// In future we may embed it in istio-agent as well, using XDS to sync the config store.
+type IstioServiceEntries struct {
```

Comment:
AFAICT there really isn't anything Istio specific about this class apart from the dependency on `model.IstioConfigStore`.  I think we could pretty easily convert this into a generic `DNSInterceptor` pretty easily.

All we would need would be to define a simple interface which would be backed by an Istio `ConfigStoreCache`:

```golang
package dns

// Just a marker interface which matches ServiceEntry's methods
type Entry interface {
  GetHosts() []string
  GetAddresses() []string
}

type  Interceptor struct {
  ...
}

func (i *Interceptor) UpdateEntries(addOrUpdate []Entry, remove []Entry) {
  ...
}
```

Then we create the `DNSInterceptor` in the server like so:

```golang
  interceptor := dns.NewInterceptor(..., initialEntries []Entry)
  configStoreCache.RegisterEventHandler(resource.GroupVersionKind {
      Group: "networking.istio.io",
      Version: "v1alpha3",
      Kind: "ServiceEntry",
    }, func(old model.Config, curr model.Config, e Event) {
      switch e {
      case model.Delete:
        entry := old.Spec.(*networking.ServiceEntry)
        interceptor.Update(nil, []dns.Entry{entry}
      default:
        entry := curr.Spec.(*networking.ServiceEntry)
        if entry.Resolution != networking.ServiceEntry_NONE {
          interceptor.Update([]dns.Entry{entry}, nil)
        }
      }
    })
```

Of course the incremental approach would mean that the DNSInterceptor would maintain its own cache of ServiceEntries. If we wanted to keep with the polling model, we could just make an interface for an `EntryProvider` that is a thin shim around the `ConfigStore`.

Reply:
Yes, that's very close to what I was planning - I wanted to reuse code that has been tested and used in the past, and get the infrastructure in place - so we can have other people work on VM and agent in parallel, while we optimize the server side.